Lipschitz-continuous functions between metric spaces (#1417)
UniMath/agda-unimath: src/metric-spaces/elements-at-bounded-distance-metric-spaces.lagda.md

# Elements at bounded distance in metric spaces

```agda
module metric-spaces.elements-at-bounded-distance-metric-spaces where
```

<details><summary>Imports</summary>

```agda
open import elementary-number-theory.positive-rational-numbers

open import foundation.dependent-pair-types
open import foundation.equivalence-relations
open import foundation.existential-quantification
open import foundation.function-types
open import foundation.functoriality-propositional-truncation
open import foundation.propositional-truncations
open import foundation.propositions
open import foundation.subtypes
open import foundation.universe-levels

open import logic.functoriality-existential-quantification

open import metric-spaces.metric-spaces

open import order-theory.preorders
```

</details>

## Idea

Two elements `x y : A` of a [metric-space](metric-spaces.metric-spaces.md) are
{{#concept "at bounded distance" Disambiguation="in a metric space" Agda=element-at-bounded-dist-Metric-Space}}
if there [exists](foundation.existential-quantification.md) some
[positive rational number](elementary-number-theory.positive-rational-numbers.md)

> `δ : ℚ⁺` such that `x` and `y` are `δ`-neighbors in `A`. Being at bounded

Reviewer: ```suggestion
`δ : ℚ⁺` such that `x` and `y` share a `δ`-neighborhood in `A`. Being at bounded
```

> distance in a metric space is an equivalence relation.

Reviewer: link equivalence relation

> distance in a metric space is an equivalence relation.

Reviewer: Maybe we could include some of our discussion on the alternative term "neighbor" here? I think we both had some good points about the term, and that could be insightful for the reader to see. If you want I can try and cook up something.

Author: If you feel like it, go ahead, but I don't think it's really necessary. I agree we both had good points for our names but you actually convinced me that `neighbor` is a bit too generic and could be misleading; this term usually carries some notion of "locality", or "small neighborhoods (like for small δ)" which isn't really reflected by this relation. 

I may come back to this topic sometimes in the future though. I think there still some interesting to say: basically a metric space is total if all elements are equivalent under this relation, so the quotient under this relation should be propositional (and contractible if the space is inhabited). In general, taking the quotient, and then subspaces of elements in a given equivalence class, we should be able to re-construct the original space as a "the total space of a family of total metrics spaces", which I think could be an interesting result, and also connects this with the other interpretation of "_total_ metric space".

## Definitions

> ### The relation of being neighbors in a metric space

Reviewer: ```suggestion
### The relation of being at bounded distance in a metric space
```

```agda
module _
  {l1 l2 : Level} (A : Metric-Space l1 l2)
  (x y : type-Metric-Space A)
  where

>   is-element-at-bounded-dist-prop-Metric-Space : Prop l2

Reviewer: Actually, the following name seems to work just fine for the relation, no?
```suggestion
  bounded-dist-prop-Metric-Space : Prop l2
```

Reviewer: or `bounded-dist-rel-prop-Metric-Space`

  is-element-at-bounded-dist-prop-Metric-Space =
    ∃ ℚ⁺ (λ d → structure-Metric-Space A d x y)

>   is-element-at-bounded-dist-Metric-Space : UU l2

Reviewer: ```suggestion
  bounded-dist-Metric-Space : UU l2
```

Reviewer: or `bounded-dist-rel-Metric-Space`

  is-element-at-bounded-dist-Metric-Space =
    type-Prop is-element-at-bounded-dist-prop-Metric-Space

  is-prop-is-element-at-bounded-dist-Metric-Space :
    is-prop is-element-at-bounded-dist-Metric-Space
  is-prop-is-element-at-bounded-dist-Metric-Space =
    is-prop-type-Prop is-element-at-bounded-dist-prop-Metric-Space
```

> ### Element-at-bounded-dists in metric spaces

Reviewer: ```suggestion
### Elements at bounded distance in metric spaces
```

```agda
module _
  {l1 l2 : Level} (A : Metric-Space l1 l2)
  (x : type-Metric-Space A)
  where

  element-at-bounded-dist-Metric-Space : UU (l1 ⊔ l2)
  element-at-bounded-dist-Metric-Space =
    type-subtype (is-element-at-bounded-dist-prop-Metric-Space A x)

  value-element-at-bounded-dist-Metric-Space :
    element-at-bounded-dist-Metric-Space → type-Metric-Space A
  value-element-at-bounded-dist-Metric-Space = pr1

  is-element-at-bounded-dist-elem-element-at-bounded-dist-Metric-Space :
    (H : element-at-bounded-dist-Metric-Space) →
    is-element-at-bounded-dist-Metric-Space
      ( A)
      ( x)
      ( value-element-at-bounded-dist-Metric-Space H)
  is-element-at-bounded-dist-elem-element-at-bounded-dist-Metric-Space = pr2
```

## Properties

> ### Being at bounded distance in a metric space is reflexive

Reviewer: ```suggestion
### The relation of being at bounded distance in a metric space is reflexive
```

```agda
module _
  {l1 l2 : Level} (A : Metric-Space l1 l2)
  where

>   refl-is-element-at-bounded-dist-Metric-Space :

Reviewer: ```suggestion
  refl-bounded-dist-Metric-Space :
```

    (x : type-Metric-Space A) → is-element-at-bounded-dist-Metric-Space A x x
  refl-is-element-at-bounded-dist-Metric-Space x =
    map-trunc-Prop
      ( λ d → d , refl-structure-Metric-Space A d x)
      ( is-inhabited-ℚ⁺)
```

### Being at bounded distance in a metric space is symmetric

```agda
module _
  {l1 l2 : Level} (A : Metric-Space l1 l2)
  where

>   symmetric-is-element-at-bounded-dist-Metric-Space :

Reviewer: ```suggestion
  symmetric-rel-bounded-dist-Metric-Space :
```

# Elements at bounded distance in metric spaces

```agda
module metric-spaces.elements-at-bounded-distance-metric-spaces where
```

<details><summary>Imports</summary>

```agda
open import elementary-number-theory.positive-rational-numbers

open import foundation.dependent-pair-types
open import foundation.equivalence-relations
open import foundation.existential-quantification
open import foundation.function-types
open import foundation.functoriality-propositional-truncation
open import foundation.propositional-truncations
open import foundation.propositions
open import foundation.subtypes
open import foundation.universe-levels

open import logic.functoriality-existential-quantification

open import metric-spaces.metric-spaces

open import order-theory.preorders
```

</details>

## Idea

Two elements `x y : A` of a [metric space](metric-spaces.metric-spaces.md) are

> {{#concept "at bounded distance" Disambiguation="in a metric space" Agda=element-at-bounded-dist-Metric-Space}}

Author: I have doubts on whether the `Agda=...` should point to `element-at-bounded-dist-Metric-Space` or `bounded-dist-Metric-Space` or both if its allowed (hence my [question](https://github.com/UniMath/agda-unimath/pull/1417/files/17c74e3ace27df5c99cf66b7dd7af5835a3ac28e#r2089445299))

Reviewer: Both is ok

Author: > Actually. Element at bounded distance is the collection of all elements at bounded distance of a given (base) point. So that should be a second concept

I replaced the link with the right definition. I'm not sure if you actually want to define a second concept for `element-at-bounded-dist-Metric-Space`  in the `## Idea` section; if so, I believe you'll write it much better than I would, so please go ahead.
